torch/jit/_recursive.py

```diff
@@ -274,6 +274,14 @@ def infer_type(name, item):
                     "to a TorchScript type.)").format(torch.typename(type(value)))
             concrete_type_builder.add_failed_attribute(name, hint)
 
+    # add hooks to concrete type
+    # TODO: would using a random number here be better? I had collisions happen
+    # even with the name and python id combination here ...  
```

Comment:
When I was writing the CPP testing code (test/jit_hooks/model.py in the top commit in this stack) I was running into hook_name + python_id collisions. Since python objects only own their id when they're in scope, the way that I was defining hooks inside of other functions was causing their python_ids to be re used. Specifically the test test_module_forward_and_pre_hook_single_IO, if you change the hook names to just be 'pre_hook' and 'forward_hook' it'll cause Python 3.8.5 to re-use the same python id with the same name pairing.

Reply:
Hm, since the goal here is to keep track of which hooks are attached to a module in order to figure out whether two modules have the same type, I wonder if you could store the actual function object in the concrete type (similar to what we do with function attributes) to get around this issue.

Reply:
Using the py::object works! It solves for the collision that was happening in the CPP testing code as well